tests-js/loan-token/BorrowingTestToken.test.js

```diff
@@ -0,0 +1,315 @@
+					"0x0" // bytes memory loanDataBytes
+				),
+				"invalid interest"
+			);
+		});
+
+		it("Test borrow insufficient collateral should fail", async () => {
+			//  prepare the test
+
+			await lend_to_pool(loanToken, SUSD, owner);
+			await set_demand_curve(loanToken);
+
+			// determine borrowing parameter
+			const withdrawAmount = oneEth.mul(new BN(10)); // i want to borrow 10 USD
+			// compute the required collateral. params: address loanToken, address collateralToken, uint256 newPrincipal,uint256 marginAmount, bool isTorqueLoan
+			let collateralTokenSent = await sovryn.getRequiredCollateral(
+				SUSD.address,
+				RBTC.address,
+				withdrawAmount,
+				new BN(10).pow(new BN(18)).mul(new BN(50)).toString(),
+				true
+			);
+			collateralTokenSent = collateralTokenSent.div(new BN(2));
+
+			//approve the transfer of the collateral
+			await RBTC.approve(loanToken.address, collateralTokenSent.toString());
+			expectRevert(
+				loanToken.borrow(
+					"0x0", // bytes32 loanId
+					withdrawAmount, // uint256 withdrawAmount
+					24 * 60 * 60, // uint256 initialLoanDuration
+					collateralTokenSent.toString(), // uint256 collateralTokenSent
+					RBTC.address, // address collateralTokenAddress
+					owner, // address borrower
+					account1, // address receiver
+					"0x0" // bytes memory loanDataBytes
+				),
+				"collateral insufficient"
+			);
+		});
+
+		it("Test borrow without early access token should fail if required", async () => {
+			// no demand curve settings -> no interest set
+			//  prepare the test
+
+			await lend_to_pool(loanToken, SUSD, owner);
+			await set_demand_curve(loanToken);
+
+			// prepare early access token
+			const early_access_token = await TestToken.new("Sovryn Early Access Token", "SEAT", 1, 10);
+			await early_access_token.transfer(account1, (await early_access_token.balanceOf(owner)).toString());
+			await loanToken.setEarlyAccessToken(early_access_token.address);
+
+			// determine borrowing parameter
+			const withdrawAmount = oneEth; // i want to borrow 10 USD
+			// compute the required collateral. params: address loanToken, address collateralToken, uint256 newPrincipal,uint256 marginAmount, bool isTorqueLoan
+			let collateralTokenSent = await sovryn.getRequiredCollateral(
+				SUSD.address,
+				RBTC.address,
+				withdrawAmount.toString(),
+				new BN(50).pow(new BN(18)),
+				true
+			);
+
+			//approve the transfer of the collateral
+			await RBTC.approve(loanToken.address, collateralTokenSent.toString());
+			expectRevert(
+				loanToken.borrow(
+					"0x0", // bytes32 loanId
+					withdrawAmount.toString(), // uint256 withdrawAmount
+					24 * 60 * 60, // uint256 initialLoanDuration
+					collateralTokenSent.toString(), // uint256 collateralTokenSent
+					RBTC.address, // address collateralTokenAddress
+					owner, // address borrower
+					account1, // address receiver
+					"0x0" // bytes memory loanDataBytes
+				),
+				"No early access tokens"
+			);
+		});
+	});
+});
```

Comment:
missing tests for the moment:
test_borrow_from_foreign_loan_should_fail
test_margin_trade_from_foreign_loan_should_fail
test_borrow_from_trade_position_should_fail

Reply:
added